src/chemistry/geoschem/cesmgc_history_mod.F90

```diff
@@ -0,0 +1,1232 @@
+!
+  SUBROUTINE CopyGCStates2Exports( am_I_Root, Input_Opt, State_Grid, HistoryConfig, LCHNK, RC )
+!
+! !USES:
+!
+  USE HCO_Interface_GC_Mod, ONLY : HCOI_GC_WriteDiagn
+  USE Input_Opt_Mod,    ONLY : OptInput
+  USE State_Grid_Mod,   ONLY : GrdState
+
+  USE cam_history,      ONLY : hist_fld_active, outfld
+!
+! !INPUT PARAMETERS:
+!
+    LOGICAL,        INTENT(IN)    :: am_I_Root
+    TYPE(OptInput), INTENT(IN)    :: Input_Opt
+    TYPE(GrdState), INTENT(IN)    :: State_Grid
+    INTEGER,        INTENT(IN)    :: LCHNK            ! Chunk number for CESM
+!
+! !INPUT AND OUTPUT PARAMETERS:
+!
+    TYPE(HistoryConfigObj), POINTER :: HistoryConfig  ! History config object
+!
+! !OUTPUT PARAMETERS:
+!
+    INTEGER,            INTENT(OUT) :: RC
+!
+! !REMARKS:
+!  !
+! !REVISION HISTORY:
+!  01 Sep 2017 - E. Lundgren - Initial version
+!  See https://github.com/geoschem/geos-chem for history
+!EOP
+!------------------------------------------------------------------------------
+!BOC
+!
+! !LOCAL VARIABLES:
+!
+    INTEGER                         :: LMAX
+    CHARACTER(LEN=255)              :: ErrMsg
+    TYPE(HistoryExportObj), POINTER :: current
+
+    ! ================================================================
+    ! CopyGCStates2Exports begins here
+    ! ================================================================
+    __Iam__('CopyGCStates2Exports (cesmgc_history_mod.F90)')
+    RC = GC_SUCCESS
+
+    ! Loop over the History Exports list
+    current => HistoryConfig%HistoryExportsList%head
+    DO WHILE ( ASSOCIATED( current ) )
+       ! Skip if not active
+       if(.not. hist_fld_active(trim(current%name))) then
+          current => current%next
+          cycle
+       endif
+
+       ! if (am_I_Root) THEN
+       !    print *, '  Copying ' // TRIM(current%name)
+       ! endif
+       IF ( current%rank == 2 ) THEN
+          IF ( ASSOCIATED( current%GCStateData2d ) ) THEN
+             current%ExportData2d(:,1:State_Grid%NY) = current%GCStateData2d(:,1:State_Grid%NY)
+          ELSE IF ( ASSOCIATED( current%GCStateData2d_4 ) ) THEN
+             current%ExportData2d(:,1:State_Grid%NY) = current%GCStateData2d_4(:,1:State_Grid%NY)
+          ELSE IF ( ASSOCIATED( current%GCStateData2d_8 ) ) THEN
+             current%ExportData2d(:,1:State_Grid%NY) = current%GCStateData2d_8(:,1:State_Grid%NY)
+          ELSE IF ( ASSOCIATED( current%GCStateData2d_I ) ) THEN
+             ! Convert integer to float (integers not allowed in MAPL exports)
+             current%ExportData2d(:,1:State_Grid%NY) = FLOAT(current%GCStateData2d_I(:,1:State_Grid%NY))
+          ELSE
+             RC = GC_FAILURE
+             ErrMsg = "No GC 2D pointer found for " // TRIM(current%name)
+             EXIT
+          ENDIF
+
+          ! Now call outfld to output for this chunk
+          call outfld(trim(current%name),                              &
+                      current%ExportData2d(1, 1:State_Grid%NY),        &   ! Chunk width always 1
```

Comment:
I'm concerned about the allocation of the ExportData pointers. For GCHP I did not need to allocate them since MAPL already does that based on what is in HISTORY.rc. I wonder if it would be better to scrap ExportData2d/3d from this file, and call outfld within each of the if associated blocks, casting to real8 in the call.